intel-sgx/async-usercalls/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+        }
+        self.core.send_usercall(usercall)
+    }
+}
+
+#[derive(Clone)]
+pub struct CallbackHandlerWaker {
+    rx: mpmc::Receiver<()>,
+    tx: mpmc::Sender<()>,
+}
+
+impl CallbackHandlerWaker {
+    fn new() -> Self {
+        let (tx, rx) = mpmc::bounded(1);
+        Self { tx, rx }
+    }
+
+    /// Interrupts the currently running or a future call to the related
+    /// CallbackHandler's `poll()`.
+    pub fn wake(&self) {
+        let _ = self.tx.try_send(());
+    }
+
+    /// Clears the effect of a previous call to `self.wake()` that is not yet
+    /// observed by `CallbackHandler::poll()`.
+    pub fn clear(&self) {
+        let _ = self.rx.try_recv();
+    }
+}
+
+pub struct CallbackHandler {
+    return_rx: mpmc::Receiver<Identified<Return>>,
+    callbacks: Mutex<HashMap<u64, Callback>>,
+    // This is used so that threads sending usercalls don't have to take the lock.
+    callback_rx: mpmc::Receiver<(u64, Callback)>,
+    waker: CallbackHandlerWaker,
+}
+
+impl CallbackHandler {
+    const RECV_BATCH_SIZE: usize = 128;
+
+    // Returns an object that can be used to interrupt a blocked `self.poll()`.
+    pub fn waker(&self) -> CallbackHandlerWaker {
+        self.waker.clone()
+    }
+
+    #[inline]
+    fn recv_returns(&self, timeout: Option<Duration>, returns: &mut [Identified<Return>]) -> usize {
+        let first = match timeout {
+            None => mpmc::select! {
+                recv(self.return_rx) -> res => res.ok(),
+                recv(self.waker.rx) -> _res => return 0,
+            },
+            Some(timeout) => mpmc::select! {
+                recv(self.return_rx) -> res => res.ok(),
+                recv(self.waker.rx) -> _res => return 0,
+                default(timeout) => return 0,
+            },
+        }
+            .expect("return channel closed unexpectedly");
+        let mut count = 0;
+        for ret in std::iter::once(first).chain(self.return_rx.try_iter().take(returns.len() - 1)) {
+            returns[count] = ret;
+            count += 1;
+        }
+        count
+    }
+
+    /// Poll for returned usercalls and execute their respective callback
+    /// functions. If `timeout` is `None`, it will block execution until at
+    /// least one return is received, otherwise it will block until there is a
+    /// return or timeout is elapsed. Returns the number of executed callbacks.
+    /// This can be interrupted using `CallbackHandlerWaker::wake()`.
+    pub fn poll(&self, timeout: Option<Duration>) -> usize {
+        // 1. wait for returns
+        let mut returns = [Identified::default(); Self::RECV_BATCH_SIZE];
+        let returns = match self.recv_returns(timeout, &mut returns) {
+            0 => return 0,
+            n => &returns[..n],
+        };
+        // 2. try to lock the mutex, if successful, receive all pending callbacks and put them in the hash map
+        let mut guard = match self.callbacks.try_lock() {
```

Comment:
It's unclear to me why the `try_lock()` is useful here. It would seem simpler, as efficient, and possibly more correct to write this as something like:
```
let mut guard = self.callbacks.lock().unwrap();
for (id, cb) in self.callback_rx.try_iter() {
     guard.insert(id, cb);
}
```

What am I missing?

Reply:
I think this is an optimization; when `self.callbacks.try_lock()` fails, it means multiple threads are competing for this resource. The code then avoids receiving more callbacks. I can imagine that `self.callback_rx.try_iter()` is a potentially time costly operation.
But I don't understand why this is correct in all cases. When the system is under very heavy load, the result of the usercall may be received in step 1 _before_ the callback is added to `self.callbacks`. This eventually leads to the callback never being called. There's also a memory leak: Eventually the callback is received in step 2, but never gets removed from the hashmap.
@mzohreva Are we missing something?

Reply:
I think you are right, this should be simplified.